dev/buildtool/versions_commands.py

```diff
@@ -0,0 +1,432 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Implements versions commands for buildtool."""
+
+import copy
+import logging
+import time
+import yaml
+
+from schema import Optional, Regex, Schema, SchemaError
+
+from buildtool import (
+    SPINNAKER_HALYARD_GCS_BUCKET_NAME,
+    SPINNAKER_CHANGELOG_BASE_URL,
+    CommandFactory,
+    CommandProcessor,
+    check_options_set,
+    check_path_exists,
+    check_subprocess,
+    raise_and_log_error,
+    write_to_path,
+)
+
+# Three stable versions are maintained at a time. https://spinnaker.io/docs/releases/versions/
+MAXIMUM_SPINNAKER_RELEASE_COUNT = 3
+
+# https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+SEMVER_REGEX = r"^(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$"
+
+# https://github.com/spinnaker/halyard/blob/d6b5ec9b98d4903824c18a603884dde0c4f4165a/halyard-core/src/main/java/com/netflix/spinnaker/halyard/core/registry/v1/Versions.java#L33
+VERSIONS_YML_SCHEMA = Schema(
+    {
+        Optional("illegalVersions"): [{"reason": str, "version": Regex(SEMVER_REGEX)}],
+        "latestHalyard": Regex(SEMVER_REGEX),
+        "latestSpinnaker": Regex(SEMVER_REGEX),
+        "versions": [
+            {
+                "alias": str,
+                "changelog": str,
+                "lastUpdate": int,
+                "minimumHalyardVersion": str,
+                "version": Regex(SEMVER_REGEX),
+            }
+        ],
+    }
+)
+
+
+def now():
+    """Hook for easier mocking."""
+    return time.time_ns() // 1_000_000
+
+
+class VersionsBuilder:
+    """Helper class for UpdateVersionsCommand that constructs the versions.yml specification."""
+
+    @staticmethod
+    def new_from_versions(versions):
+        """Create a new VersionsBuilder with the supplied versions.yml as a dict."""
+        return VersionsBuilder(base_versions=versions)
+
+    @property
+    def base_versions(self):
+        return self.__base_versions_dict
+
+    def __init__(self, base_versions=None):
+        """Construct new builder.
+
+        Args:
+          base_versions[dict]: If defined, this is a versions dict to start with.
+                          It is intended to support an "update" use case where
+                          only a subset of entries are updated within it.
+        """
+        self.__base_versions_dict = base_versions or {
```

Comment:
I could use a comment about why this default value is helpful.  My reflex is to ditch the default and require `base_versions`, but I bet it's useful.

Reply:
I agree that it seems redundant to our general use case of updating the public GCS versions.yml file.
For extra context:

The `BomBuilder` and others all have a default dict which I have followed and rightly/wrongly it is easier than dealing with some none/nil/undefined behaviours.
I did miss the corresponding test case and have added in https://github.com/spinnaker/buildtool/pull/195/commits/87b42c2ca9f69c83f38fc33060d9c5bd39f1b2f8

That said, the `./buildtool.sh update_versions` command enforces passing in a `versions.yml` path which I thought was safer for anyone invoking buildtool. See: https://github.com/spinnaker/buildtool/pull/195/files#diff-9b53a58ad4a894dad2080a292e688a93b2f02c1fd93ed1c12043e6e0ee8fd63fR312.
So... there's a mismatch.

Reply:
Fair enough.  Let's ship it.